File: leetcode/scala/MergeSortedArray.scala
object MergeSortedArray {
  // Leetcode 88: https://leetcode.com/problems/merge-sorted-array/
  def merge(nums1: Array[Int], m: Int, nums2: Array[Int], n: Int): Unit = {
    var curr = m + n - 1
    var (i, j) = (m - 1, n - 1)

    while (i >= 0 && j >= 0) {
      if (nums1(i) < nums2(j)) {
        nums1(curr) = nums2(j)
        j -= 1
      } else {
        nums1(curr) = nums1(i)
        i -= 1
      }
      curr -= 1
    }

    while (j >= 0) {
      nums1(curr) = nums2(j)
      j -= 1
      curr -= 1
    }
  }
}

File: leetcode/python/merge_sorted_array.py
from typing import List

# Leetcode 88: https://leetcode.com/problems/merge-sorted-array/


def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
    """
    Do not return anything, modify nums1 in-place instead.
    """
    curr = m + n - 1
    m -= 1
    n -= 1

    while m >= 0 and n >= 0:
        if nums2[n] > nums1[m]:
            nums1[curr] = nums2[n]
            n -= 1
        else:
            nums1[curr] = nums1[m]
            m -= 1
        curr -= 1

    if n >= 0:
        nums1[0:n+1] = nums2[0:n+1]
